pkg/cluster/sync.go

```diff
@@ -107,73 +78,159 @@ func (c *Cluster) Sync(newSpec *spec.Postgresql) (err error) {
 	return
 }
 
+func (c *Cluster) syncServices() error {
+	for _, role := range []PostgresRole{Master, Replica} {
+		c.logger.Debugf("syncing %s service", role)
+
+		if err := c.syncEndpoint(role); err != nil {
+			return fmt.Errorf("could not sync %s endpont: %v", role, err)
+		}
+
+		if err := c.syncService(role); err != nil {
+			return fmt.Errorf("could not sync %s service: %v", role, err)
+		}
+	}
+
+	return nil
+}
+
 func (c *Cluster) syncService(role PostgresRole) error {
-	cSpec := c.Spec
-	if c.Services[role] == nil {
-		c.logger.Infof("could not find the cluster's %s service", role)
-		svc, err := c.createService(role)
-		if err != nil {
-			return fmt.Errorf("could not create missing %s service: %v", role, err)
+	c.setProcessName("syncing %s service", role)
+
+	svc, err := c.KubeClient.Services(c.Namespace).Get(c.serviceName(role), metav1.GetOptions{})
+	if err == nil {
+		if role == Replica && !c.Spec.ReplicaLoadBalancer {
+			if err := c.deleteService(role); err != nil {
+				return fmt.Errorf("could not delete %s service", role)
+			}
 		}
-		c.logger.Infof("created missing %s service %q", role, util.NameFromMeta(svc.ObjectMeta))
+
+		desiredSvc := c.generateService(role, &c.Spec)
+		match, reason := k8sutil.SameService(svc, desiredSvc)
+		if match {
+			c.Services[role] = svc
+			return nil
+		}
+		c.logServiceChanges(role, svc, desiredSvc, false, reason)
+
+		if err := c.updateService(role, desiredSvc); err != nil {
+			return fmt.Errorf("could not update %s service to match desired state: %v", role, err)
+		}
+		c.logger.Infof("%s service %q is in the desired state now", role, util.NameFromMeta(desiredSvc.ObjectMeta))
 
 		return nil
+	} else if !k8sutil.ResourceNotFound(err) {
+		return fmt.Errorf("could not get %s service: %v", role, err)
 	}
+	c.Services[role] = nil
 
-	desiredSvc := c.generateService(role, &cSpec)
-	match, reason := c.sameServiceWith(role, desiredSvc)
-	if match {
+	// Service does not exist
+	if role == Replica && !c.Spec.ReplicaLoadBalancer {
 		return nil
 	}
-	c.logServiceChanges(role, c.Services[role], desiredSvc, false, reason)
 
-	if err := c.updateService(role, desiredSvc); err != nil {
-		return fmt.Errorf("could not update %s service to match desired state: %v", role, err)
+	c.logger.Infof("could not find the cluster's %s service", role)
+
+	if svc, err := c.createService(role); err != nil {
+		if k8sutil.ResourceAlreadyExists(err) {
+			c.logger.Infof("%s service %q already exists", role, util.NameFromMeta(svc.ObjectMeta))
```

Comment:
in this case we won't reach the outer else block, that sets c.Services[role] = sec, so c.Services[role] will be nil despite the fact that the actual service exists.